File: April/Abiji-2020/Golang /Smallest String starting from Leaf/smallest string starting from leaf.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func smallestFromLeaf(root *TreeNode) string {
    if root == nil {
        return ""
    }
    
    // Define a helper function for DFS traversal
    var dfs func(node *TreeNode, path string, smallest *string)
    dfs = func(node *TreeNode, path string, smallest *string) {
        if node == nil {
            return
        }
        
        // Concatenate the current character to the path
        path = string('a' + byte(node.Val)) + path
        
        // If it's a leaf node, compare the path with the smallest string found so far
        if node.Left == nil && node.Right == nil {
            if *smallest == "" || path < *smallest {
                *smallest = path
            }
            return
        }
        
        // Recur for left and right subtrees
        dfs(node.Left, path, smallest)
        dfs(node.Right, path, smallest)
    }
    
    // Initialize the smallest string as an empty string
    smallest := ""
    
    // Start DFS traversal from the root node with an empty path
    dfs(root, "", &smallest)
    
    return smallest
}

File: April/Abiji-2020/C++/Smallest String Starting From Leaf/smallest string starting from leaf.cpp
#include <algorithm>
#include <string>
using namespace std;

class Solution {
    string dfs(TreeNode* node, string path, string& smallest) {
        if (!node) return ""; // Base case: if the node is NULL, return an empty string
        path += char('a' + node->val);
        if (!node->left && !node->right) { // If it's a leaf node
            reverse(path.begin(), path.end());
            if (smallest.empty() || smallest > path) {
                smallest = path;
            }
            return smallest; // Return smallest after processing leaf node
        }
        // Recur for left and right subtrees
        dfs(node->left, path, smallest);
        dfs(node->right, path, smallest);
        return smallest; // Return smallest from current subtree
    }
public:
    string smallestFromLeaf(TreeNode* root) {
        string smallest;
        smallest = dfs(root, "", smallest);
        return smallest;
    }
};
